### PR TITLE
Enable nestmate J2I thunk AOT support

### PR DESCRIPTION
**Pull request by NigelYiboYu**
- Enable J2I thunk relocation for nestmate

- Replace addProjectSpecializedRelocation API with
  addExternalRelocation() API, which is native to the OpenJ9 project.

Signed-off-by: Nigel Yu <yunigel@ca.ibm.com>

### CODE REVIEW

**Comment by NigelYiboYu** [comment]
Marking as WIP until all tests pass.

**Comment by NigelYiboYu** [comment]
This pull request could fix a JDK11 failure: 
https://ci.eclipse.org/openj9/job/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJDK11/3/tapResults/

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp [hunk 1]
```diff
@@ -501,6 +501,9 @@ TR::S390VirtualUnresolvedSnippet::emitSnippetBody()
 
    cursor = generatePICBinary(cg(), cursor, glueRef);
 
+   // J2I relocation information for private nestmate calls
+   auto j2iRelocInfo = reinterpret_cast<TR_RelocationRecordInformation*>(comp->trMemory()->allocateMemory(sizeof(TR_RelocationRecordInformation), heapAlloc));
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: Can we please move this declaration as close as possible to where the data is actually initialized. No reason for it to live so far away making it's scope huge.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp [hunk 2]
```diff
@@ -515,12 +518,17 @@ TR::S390VirtualUnresolvedSnippet::emitSnippetBody()
...
 
+   // CP addr
    *(uintptrj_t *) cursor = (uintptrj_t) callNode->getSymbolReference()->getOwningMethod(comp)->constantPool();
 
-   AOTcgDiag1(comp, "add TR_Thunks cursor=%x\n", cursor);
-   cg()->addExternalRelocation(new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor, *(uint8_t **)cursor, getNode() ? (uint8_t *)(intptr_t)getNode()->getInlinedSiteIndex() : (uint8_t *)-1, TR_Thunks, cg()),
-                             __FILE__, __LINE__, callNode);
+   AOTcgDiag1(comp, "add TR_J2IVirtualThunkPointer cursor=%x\n", cursor);
+   cg()->addExternalRelocation(new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor, (uint8_t*)j2iRelocInfo, NULL,
+                                                                                 TR_J2IVirtualThunkPointer, cg()),
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: I'd prefer if we initialize `j2iRelocInfo` before passing it as an argument here. What if `TR::ExternalRelocation` were to cache some of the results from this argument? We would be reading garbage memory then. This relocation should be moved to after `data3` is initialized.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp [hunk 3]
```diff
@@ -1026,22 +1032,32 @@ TR::J9S390InterfaceCallDataSnippet::emitSnippetBody()
    getSnippetLabel()->setCodeLocation(cursor);
    TR::Node * callNode = getNode();
 
+   // J2I relocation information for private nestmate calls
+   auto j2iRelocInfo = reinterpret_cast<TR_RelocationRecordInformation*>(comp->trMemory()->allocateMemory(sizeof(TR_RelocationRecordInformation), heapAlloc));
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: All similar points to above need to be addressed for this one as well.

**Comment by fjeremic** [comment]
See Power implementation for reference:
https://github.com/eclipse/openj9/pull/2882

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp (hunk 3 above)
Comment (NigelYiboYu) [reply runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: Variable scope reduced.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp (hunk 1 above)
Comment (NigelYiboYu) [reply runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: Variable scope reduced.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp (hunk 2 above)
Comment (NigelYiboYu) [reply runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: The `j2iRelocInfo` is now fully initialized before used.

**Comment by irinarada** [comment]
Thanks @NigelYiboYu . @pshipton FYI.

**Comment by irinarada** [comment]
Other AOT failures: 
https://ci.eclipse.org/openj9/job/Test-sanity.functional-JDK11-linux_ppc-64_cmprssptrs_le/19/tapResults/
https://ci.eclipse.org/openj9/job/Test-sanity.functional-JDK11-win_x86-64_cmprssptrs/17/tapResults/

@NigelYiboYu - can you check if they're fixed with everything is in so far. Wondering if they're related to @jdmpapin's code instead as they're Win and PPC (not z).

**Comment by fjeremic** [comment]
Jenkins test sanity zlinux JDK8,JDK11

**Comment by fjeremic** [comment]
JDK11 failures in above build due to #2890 for which a fix was merged. I'm happy to merge this as soon as WIP is removed whenever @NigelYiboYu's testing is done.

**Comment by fjeremic** [comment]
Travis failed due to:

```
The job exceeded the maximum time limit for jobs, and has been terminated.
```

Ignoring this build.

**Comment by NigelYiboYu** [comment]
JDK11 failures below dont' seem to be related to AOT:
https://ci.eclipse.org/openj9/job/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJ9/24/tapResults/

The two failed tests are `J9vmTest_4` and `J9vmTest_5`
Failure messages are:

```
+++ j9vm.test.clone.CloneWeakReferenceTest: +++
command: /home/jenkins/jenkins-agent/workspace/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJ9/build/linux-s390x-normal-server-release/images/jdk/bin/java  -Xcompressedrefs -XX:+UseCompressedOops -Xjit -Xgcpolicy:balanced  -Xdump -Xms64m -Xmx64m    -classpath /home/jenkins/jenkins-agent/workspace/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/functional/VM_Test/VM_Test.jar:/home/jenkins/jenkins-agent/workspace/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/TestConfig/lib/asm-all.jar  j9vm.test.clone.CloneWeakReferenceTest 

Exception in thread "main" java.lang.CloneNotSupportedException: Create a new Reference, since a Reference cannot be cloned.
	at java.lang.ref.Reference.clone(java.base@11-internal/Reference.java:221)
	at j9vm.test.clone.CloneWeakReferenceTest$SubWeakReference.clone(CloneWeakReferenceTest.java:85)
	at j9vm.test.clone.CloneWeakReferenceTest.main(CloneWeakReferenceTest.java:44)
no-zero exit value: 1
*** Test FAILED *** (j9vm.test.clone.CloneWeakReferenceTest)
```

**Pull request merged**